core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java

```diff
@@ -0,0 +1,513 @@
+          filesGBL = getFilesGroupedByLocation(fileSelector);
+          UtilWaitThread.sleep(250);
+        } else {
+          break;
+        }
+      }
+    } finally {
+      execSrv.shutdownNow();
+    }
+
+    return partitionSummaries;
+  }
+
+  public static interface FileSystemResolver {
+    FileSystem get(Path file);
+  }
+
+  /**
+   * This method will read summaries from a set of files.
+   */
+  public SummaryCollection processFiles(FileSystemResolver volMgr, Map<String,List<TRowRange>> files, BlockCache summaryCache, BlockCache indexCache,
+      ExecutorService srp) {
+    SummaryCollection fileSummaries = new SummaryCollection();
+
+    List<Future<SummaryCollection>> futures = new ArrayList<>();
+    for (Entry<String,List<TRowRange>> entry : files.entrySet()) {
+      Future<SummaryCollection> future = srp.submit(() -> {
+        List<RowRange> rrl = Lists.transform(entry.getValue(), RowRange::new);
+        return getSummaries(volMgr, entry.getKey(), rrl, summaryCache, indexCache);
+      });
+      futures.add(future);
+    }
+
+    for (Future<SummaryCollection> future : futures) {
+      try {
+        fileSummaries.merge(future.get(), factory);
+      } catch (ExecutionException | InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    return fileSummaries;
+  }
+
+  private int countFiles() throws TableNotFoundException, AccumuloException, AccumuloSecurityException {
+    // TODO use a batch scanner + iterator to parallelize counting files
+    Iterable<TabletMetadata> tmi = MetadataScanner.builder().from(ctx).overUserTableId(tableId, startRow, endRow).fetchFiles().fetchPrev().build();
+    return (int) StreamSupport.stream(tmi.spliterator(), false).mapToInt(tm -> tm.getFiles().size()).sum();
+  }
+
+  private class GatherRequest implements Callable<SummaryCollection> {
+
+    private int remainder;
+    private int modulus;
+    private TInfo tinfo;
+
+    GatherRequest(TInfo tinfo, int remainder, int modulus) {
+      this.remainder = remainder;
+      this.modulus = modulus;
+      this.tinfo = tinfo;
+    }
+
+    @Override
+    public SummaryCollection call() throws Exception {
+      TSummaryRequest req = getRequest();
+
+      ClientContext cct = new ClientContext(ctx.getInstance(), ctx.getCredentials(), ctx.getConfiguration()) {
+        @Override
+        public long getClientTimeoutInMillis() {
+          return Math.max(super.getClientTimeoutInMillis(), 10 * 60 * 1000);
+        }
+      };
+
+      TSummaries tSums = ServerClient.execute(cct, c -> c.getSummariesForPartition(tinfo, ctx.rpcCreds(), req, modulus, remainder));
+      return new SummaryCollection(tSums);
+    }
+  }
+
+  public SummaryCollection gather() throws TableNotFoundException, AccumuloException, AccumuloSecurityException {
+
+    ExecutorService es = Executors.newFixedThreadPool(ctx.getConnector().instanceOperations().getTabletServers().size());
```

Comment:
Yeah this not ideal.  These threads will not consume CPU, they are just waiting on remote operations.  The threads will consume memory resources though.  One nice solution for this would be to make all of the request asynchronously AND not use threads.  I looked into doing this, it would be a lot of new code that would need to be tested for fault tolerance.  I will open a follow up issue for this.